lib/timeout.rb

```diff
@@ -25,27 +25,31 @@
 module Timeout
   VERSION = "0.3.2"
 
+  # Internal error raised to when a timeout is triggered.
+  class ExitException < Exception
+    # Return the receiver if it the exception is triggered in
+    # the same thread.
+    def exception(*a)
+      if @thread == Thread.current
+        self
+      else
+        super
+      end
+    end
+  end
+
   # Raised by Timeout.timeout when the block times out.
   class Error < RuntimeError
-    attr_reader :thread
-
-    def self.catch(*args)
-      exc = new(*args)
+    def self.handle_timeout(message)
+      exc = ExitException.new(message)
       exc.instance_variable_set(:@thread, Thread.current)
-      exc.instance_variable_set(:@catch_value, exc)
-      ::Kernel.catch(exc) {yield exc}
-    end
 
-    def exception(*)
-      # TODO: use Fiber.current to see if self can be thrown
-      if self.thread == Thread.current
-        bt = caller
-        begin
-          throw(@catch_value, bt)
-        rescue UncaughtThrowError
-        end
+      begin
+        yield exc
+      rescue ExitException => e
+        raise new(message) unless exc.equal?(e)
+        super
```

Comment:
What does `super` do here?

Reply:
That should be fixed.  It should `raise` instead.  I'll push a fix.  This is necessary for some nested timeouts, so only convert from Timeout::ExitException to Timeout::Error for the matching Timeout.timeout call.

Reply:
Yep, I added a nested timeouts test which handles this case.